### ISSUE TITLE
custom Function intermediary results are dangerous for double backward

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#58509]
When using custom autograd.Function, it is easy to save an intermediary result in the context like `ctx.foo` and then use it during the backward pass.
Unfortunately, if that value is actually linked to the input/output gradient computation, the double backward computation will be wrong because the contribution of `ctx.foo` will be ignored.

We can reduce this risk by intercepting all Tensors that are saved onto `ctx.` during the forward pass and attaching to them a grad_fn that will fail during the double backward pass if this Tensor was used during that computation in a differentiable manner.
The error should suggest to the user to make such intermediary result an `output` of the function so that it will have a proper `grad_fn` associated with it.

A bad example is:
```python
import torch
from torch import autograd

class BadCustomFunction(autograd.Function):
  @staticmethod
  def forward(ctx, inp):
    interm = inp * 2
    ctx.foo = interm
    res = interm ** 2
    return res

  @staticmethod
  def backward(ctx, gres):
      grad = 2 * 2 * ctx.foo * gres
      return grad


inp = torch.rand(2, dtype=torch.double, requires_grad=True)
# Gradcheck is correct
autograd.gradcheck(BadCustomFunction.apply, inp)  # True
# Double grad is silently wrong
autograd.gradgradcheck(BadCustomFunction.apply, inp)  # False

```

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

**Comment by soulitzer** [comment]
Closing in favor of https://github.com/pytorch/pytorch/issues/76185
